backup/server.py

```diff
@@ -0,0 +1,87 @@
+        self.bind.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        self.bind.bind(addr)
+
+    def _send_packet(self, typ: int, payload: bytes) -> None:
+        send_packet(self.sock, typ, payload)
+
+    def _recv_packet(self) -> Tuple[int, bytes]:
+        return recv_packet(self.sock)
+
+    def _handle_conn(self, conn) -> None:
+        # Can only handle one connection at a time
+        logging.info('Servicing incoming connection')
+        self.sock = conn
+        while True:
+            try:
+                (typ, payload) = self._recv_packet()
+            except IOError as e:
+                logging.info('Connection closed')
+                break
+            if typ in PKT_CHANGE_TYPES:
+                change = change_from_packet(typ, payload)
+                if typ == PacketType.CHANGE:
+                    logging.info('Received CHANGE {}'.format(change.version))
+                else:
+                    logging.info('Received SNAPSHOT {}'.format(change.version))
+                self.backend.add_change(change)
+                self._send_packet(PacketType.ACK, struct.pack("!I", self.backend.version))
+            elif typ == PacketType.REWIND:
+                logging.info('Received REWIND')
+                to_version, = struct.unpack('!I', payload)
+                if to_version != self.backend.prev_version:
+                    logging.info('Cannot rewind to version {}'.format(to_version))
+                    self._send_packet(PacketType.NACK, struct.pack("!I", self.backend.version))
+                else:
+                    self.backend.rewind()
+                    self._send_packet(PacketType.ACK, struct.pack("!I", self.backend.version))
+            elif typ == PacketType.REQ_METADATA:
+                logging.info('Received REQ_METADATA')
+                blob = struct.pack("!IIIQ", 0x01, self.backend.version,
+                           self.backend.prev_version,
+                           self.backend.version_count)
+                self._send_packet(PacketType.METADATA, blob)
+            elif typ == PacketType.RESTORE:
+                logging.info('Received RESTORE')
+                for change in self.backend.stream_changes():
+                    (typ, payload) = packet_from_change(change)
+                    self._send_packet(typ, payload)
+                self._send_packet(PacketType.DONE, b'')
+            elif typ == PacketType.COMPACT:
+                logging.info('Received COMPACT')
+                stats = self.backend.compact()
+                self._send_packet(PacketType.COMPACT_RES, json.dumps(stats).encode())
+            elif typ == PacketType.ACK:
+                logging.info('Received ACK')
+            elif typ == PacketType.NACK:
+                logging.info('Received NACK')
+            elif typ == PacketType.METADATA:
+                logging.info('Received METADATA')
+            elif typ == PacketType.COMPACT_RES:
+                logging.info('Received COMPACT_RES')
```

Comment:
Aren't these types unexpected to the server as well?

Reply:
Yes, true.

Reply:
Also, the client is [allowed to ACK](https://github.com/lightningd/plugins/blob/62745573b21a2cb165895f507a7236397afd9dff/backup/remote.md#restore) changes from the server when retrieving the backup. It currently doesn't.

I guess a NACK would be fatal though. It should be handled in some way that is not "unexpected message".

I would prefer to leave mailing this down for later.